### PR TITLE
Feature/improve feedstate logic

### LINKED ISSUE #166: Actions should be removed/cancelled on disconnect, deactivation & catalog disable
### Describe the bug:
There's a scheduled action (`WooCommerce > Status > Scheduled Actions`) for registering and generating the feed file (and maybe other feed stuff).

If the merchant disconnects from Pinterest, all feed-related code should stop (as I understand it). However, the action continues to run:

Screen Shot 2021-08-19 at 10 01 49 AM

Note also there are two actions in the table. Is this intended? Seems like a bug.

### Steps to reproduce:
1. Onboard and enable product sync.
2. View the `Pending` actions in `WooCommerce > Status > Scheduled Actions` and search for `pinterest`.
3. Disconnect from Pinterest.

#### Expected behavior
- The actions should be removed/cancelled if user does any of the following:
  - disconnects from Pinterest
  - deactivates or uninstalls Pinterest plugin
  - disables catalog sync.
- Ideally only one scheduled/recurring action. If two are needed, give them different hook names and make it clear how they relate.

In general, if catalog sync is not active, no code relating to it should run.

#### Actual behavior
- 2x recurring actions that look the same.
- Actions continue to run when user has disconnected their store from Pinterest.

### LINKED ISSUE #178: Do not store progress as translated strings in the database
While working on #158, it became relevant that there's strings that are stored in the database to mark things like progress. These are stored translated (and after changes implemented in #163) with HTML markup even.

One particular example is the progress status for the feed generation (Wrote %s out of %s to file). 

I'm opening up this issue by request of @haszari to reinforce this:

- Markup / UI text should be generated as late as possible. 
- State should not be cached unless there's a specific need. If stats/data needs to be stored  for later display in UI (e.g. num products in feed so far, feed generation time, feed file size), store in transients so it's clear it's derived, temporary data. Store raw values, not generated markup. Prefer one value per transient, rather than storing a large nested structure.

### PR DESCRIPTION
**Pull request by ksere**
Closes #178 
Closes #166 

### Changes proposed in this pull request

Small-ish refactor of the `ProductSync` methods related to handling the status of product feed.
More specifically;

- Improves logic of how we set & retrieve the status of the product feed by refactoring `ProductSync::feed_job_status()` onto smaller get/set methods. 
- Simplifies/reduces stored data for feed's current status.
- Uses transients for saving the related data.
- Avoids storing progress as string in the database.  
- Moves deregistration methods to concurrent request
- Avoid unnecessary calls to reschedule.
- Re-implements cleaner `reschedule_if_expired()`
- Implements `reschedule_if_errored()` to restart job after a delay.
- Remove scheduled actions on plugin deactivation. 

#### Screenshots

No visible changes to user flows / interactions. 

### Detailed test instructions

1. Start with a clean plugin installation and a new Pinterest account.
2. Start the onboarding process and Connect to Pinterest.
3. Complete onboarding process
4. Verify that the product feed is being generated properly by visiting the Catalog tab of the plugin.
5. Verify that the product feed is being registered with Pinterest properly using Pinterest's dashboard.
5. Verify that the product feed is being regenerated after it expires (24h)
6. Verify that progress is not stored as translated strings as stated in #178
6. Verify that when Product Sync is disabled, the feed file is being deleted, the related transients are being removed and the scheduled actions are being removed. 
7. Verify that when the plugin is deactivated, scheduled tasks are removed.

### CODE REVIEW

File: src/API/FeedState.php [hunk 1]
```diff
@@ -189,9 +192,12 @@ private function add_local_feed_state( $result ) {
 				$status_label = esc_html__( 'Up to date', 'pinterest-for-woocommerce' );
 				$extra_info   = sprintf(
 					/* Translators: %1$s Time string, %2$s text string indicating progress */
```
Comment (msaggiorato) [inline src/API/FeedState.php]: Will need to update translators string here.

File: src/API/FeedState.php [hunk 2]
```diff
@@ -178,9 +178,12 @@ private function add_local_feed_state( $result ) {
 				$status_label = esc_html__( 'Feed generation in progress.', 'pinterest-for-woocommerce' );
 				$extra_info   = sprintf(
 					/* Translators: %1$s Time string, %2$s text string indicating progress */
```
Comment (msaggiorato) [inline src/API/FeedState.php]: Will need to update translators string here.

File: src/API/FeedState.php [hunk 3]
```diff
@@ -212,7 +218,7 @@ private function add_local_feed_state( $result ) {
...
```
Comment (msaggiorato) [inline src/API/FeedState.php]: Will need to update translators string here.

**Review by msaggiorato — CHANGES_REQUESTED** [review]
I'd say this is good to go to QA, just need to update some translateable strings.

**Comment by budzanowski** [comment]
This looks good to go.

**Review by budzanowski — APPROVED** [review]

**Review by msaggiorato — APPROVED** [review]

**Pull request merged**